lighthouse/utils/runner.py

```diff
@@ -0,0 +1,166 @@
+"""
+Utility functions for running workloads.
+"""
+
+import numpy as np
+import os
+from mlir import ir
+from mlir.dialects import func, arith, scf, memref
+from mlir.execution_engine import ExecutionEngine
+from mlir.runtime.np_to_memref import get_ranked_memref_descriptor
+from lighthouse.utils.mlir import get_mlir_library_path
+from lighthouse.utils import memrefs_to_packed_args
+from lighthouse import Workload
+from typing import Optional
+
+
+def get_engine(
+    payload_module: ir.Module, shared_libs: list[str] = None, opt_level: int = 3
+) -> ExecutionEngine:
+    lib_dir = get_mlir_library_path()
+    libs = []
+    for so_file in shared_libs or []:
+        so_path = os.path.join(lib_dir, so_file)
+        if not os.path.isfile(so_path):
+            raise ValueError(f"Could not find shared library {so_path}")
+        libs.append(so_path)
+    execution_engine = ExecutionEngine(
+        payload_module, opt_level=opt_level, shared_libs=libs
+    )
+    execution_engine.initialize()
+    return execution_engine
+
+
+def execute(
+    workload: Workload,
+    check_correctness: bool = True,
+    schedule_parameters: Optional[dict] = None,
+    verbose: int = 0,
+):
+    # lower payload with schedule
+    payload_module = workload.lower_payload(schedule_parameters=schedule_parameters)
+    # get execution engine
+    engine = get_engine(payload_module, shared_libs=workload.shared_libs())
+
+    with workload.allocate_inputs(execution_engine=engine) as inputs:
+        # prepare function arguments
+        packed_args = memrefs_to_packed_args(inputs)
+
+        # handle to payload function
+        payload_func = engine.lookup(workload.payload_function_name)
+
+        # call function
+        payload_func(packed_args)
+
+        if check_correctness:
+            success = workload.check_correctness(
+                execution_engine=engine, verbose=verbose
+            )
+            if not success:
+                raise ValueError("Benchmark verification failed.")
+
+
+def emit_benchmark_function(
```

Comment:
This function seems to be largely independent of the `workload` argument.

I would probably still prefer it in `lighthouse.workload` over `lighthouse.utils` though.

Reply:
Updated now so that this function does not depend on workload at all. Moved to lighthouse/workload with runner.py